package/network/utils/umbim/files/lib/netifd/proto/mbim.sh

```diff
@@ -84,93 +89,143 @@ _proto_mbim_setup() {
 		proto_notify_error "$interface" NO_REGISTRATION
 		return 1
 	}
 	tid=$((tid + 1))
 
 	echo "mbim[$$]" "Attach to network"
-   	umbim $DBG -n -t $tid -d $device attach || {
+	umbim $DBG -n -t $tid -d $device attach || {
 		echo "mbim[$$]" "Failed to attach to network"
 		proto_notify_error "$interface" ATTACH_FAILED
 		return 1
 	}
 	tid=$((tid + 1))
- 
+
 	echo "mbim[$$]" "Connect to network"
-	while ! umbim $DBG -n -t $tid -d $device connect "$apn" "$auth" "$username" "$password"; do
+	while ! umbim $DBG -n -t $tid -d $device connect "$pdptype:$apn" "$auth" "$username" "$password"; do
 		tid=$((tid + 1))
 		sleep 1;
 	done
 	tid=$((tid + 1))
 
+	echo "mbim[$$]" "Connected, obtain IP address and configure interface"
+	config="/tmp/mbim.$$.config"
+	umbim $DBG -n -t $tid -d $device config > "$config" || {
+		echo "mbim[$$]" "Failed to obtain IP address"
+		proto_notify_error "$interface" CONFIG_FAILED
+		rm -f "$config"
+		return 1
+	}
+	cat "$config"
+	tid=$((tid + 1))
+
 	uci_set_state network $interface tid "$tid"
 
-	echo "mbim[$$]" "Connected, starting DHCP"
+	local ip_4 ip_6
+	ip_4=$(awk '$1=="ipv4address:" {print $2}' "$config")
+	ip_6=$(awk '$1=="ipv6address:" {print $2}' "$config")
+	[ "$ip_4" ] || [ "$ip_6" ] || {
+		echo "mbim[$$]" "Failed to obtain IP addresses"
+		proto_notify_error "$interface" CONFIG_FAILED
+		rm -f "$config"
+		return 1
+	}
+
 	proto_init_update "$ifname" 1
-	proto_send_update "$interface"
+	proto_set_keep 1
+	local ip mask gateway mtu dns1 dns2
+
+	[ "$ip_4" ] && {
+		echo "mbim[$$]" "Configure IPv4 on $ifname"
+		ip=${ip_4%%/*}
+		mask=${ip_4##*/}
+		gateway=$(awk '$1=="ipv4gateway:" {print $2; exit}' "$config")
+		mtu=$(awk '$1=="ipv4mtu:" {print $2; exit}' "$config")
+		dns1=$(awk '$1=="ipv4dnsserver:" {print $2}' "$config" | sed -n "1p")
+		dns2=$(awk '$1=="ipv4dnsserver:" {print $2}' "$config" | sed -n "2p")
+
+		proto_add_ipv4_address "$ip" "$mask"
+		[ "$defaultroute" = 0  ] || proto_add_ipv4_route 0.0.0.0 0 "$gateway" "$ip_4" "$metric"
+		[ "$peerdns" = 0 ] || {
+			[ "$dns1" ] && proto_add_dns_server "$dns1"
+			[ "$dns2" ] && proto_add_dns_server "$dns2"
+		}
+		[ "$mtu" ] && ip link set "$ifname" mtu "$mtu"
+	}
+	[ "$ip_6" ] && {
+		echo "mbim[$$]" "Configure IPv6 on $ifname"
+		ip=${ip_6%%/*}
+		mask=${ip_6##*/}
+		gateway=$(awk '$1=="ipv6gateway:" {print $2; exit}' "$config")
+		mtu=$(awk '$1=="ipv6mtu:" {print $2; exit}' "$config")
+		dns1=$(awk '$1=="ipv6dnsserver:" {print $2}' "$config" | sed -n "1p")
+		dns2=$(awk '$1=="ipv6dnsserver:" {print $2}' "$config" | sed -n "2p")
+
+		proto_add_ipv6_address "$ip" "$mask"
+		proto_add_ipv6_prefix "$ip_6"
+		[ "$defaultroute" = 0  ] || proto_add_ipv6_route "::" 0 "$gateway" "" "" "$ip_6" "$metric"
+		[ "$peerdns" = 0 ] || {
+			[ "$dns1" ] && proto_add_dns_server "$dns1"
+			[ "$dns2" ] && proto_add_dns_server "$dns2"
+		}
+		[ "$mtu" ] && ip link set "$ifname" mtu "$mtu"
+		# TODO: Don't know how to propagate IPv6 address to lan from here?
```

Comment:
You cannot propagate the interface IPv6, you need a separate routed prefix for that. Not sure if such mobile connections support requesting IPv6 prefixes via DHCPv6-PD.

Reply:
DHCPv6-PD is possible, but I've never seen it implemented by any operator.  So I believe we can ignore it for now.

But you can propagate the /64 you are assigned. The mobile interface is not a LAN interface, even if the drivers or modem firmware make it look like one.  There is no on-link prefix.  The assigned /64 is a routed prefix. There is no "interface address".  The modem firmware will just print a random address every time you query it.  You  can safely ignore it.  Steal a single /128 for the OpenWrt router, if you need IPv6 connectivity to it, and route the rest of the /64 wherever you need it.

Notice how the modem firmware gives a new address for every query, despite the connection and prefix being the same (this is the EM7455 in my laptop, but that's irrelevant):


```
root@miraculix:/home/bjorn# umbim -p -d /dev/cdc-wdm0 config 
  ipv4address: 10.197.125.191/25
  ipv4gateway: 10.197.125.192
  ipv4mtu: 1500
  ipv4dnsserver: 193.213.112.4
  ipv4dnsserver: 130.67.15.198
  ipv6address: 2a02:2121:288:ef9d:e832:4b63:ea3:ad82/64
  ipv6gateway: 2a02:2121:288:ef9d:f446:f8f3:886b:97d0
  ipv6mtu: 1500
  ipv6dnsserver: 2001:4600:4:fff:0:0:0:52
  ipv6dnsserver: 2001:4600:4:1fff:0:0:0:52
root@miraculix:/home/bjorn# umbim -p -d /dev/cdc-wdm0 config 
  ipv4address: 10.197.125.191/25
  ipv4gateway: 10.197.125.192
  ipv4mtu: 1500
  ipv4dnsserver: 193.213.112.4
  ipv4dnsserver: 130.67.15.198
  ipv6address: 2a02:2121:288:ef9d:195b:a6ca:6d1d:f9c/64
  ipv6gateway: 2a02:2121:288:ef9d:f446:f8f3:886b:97d0
  ipv6mtu: 1500
  ipv6dnsserver: 2001:4600:4:fff:0:0:0:52
  ipv6dnsserver: 2001:4600:4:1fff:0:0:0:52
root@miraculix:/home/bjorn# umbim -p -d /dev/cdc-wdm0 config 
  ipv4address: 10.197.125.191/25
  ipv4gateway: 10.197.125.192
  ipv4mtu: 1500
  ipv4dnsserver: 193.213.112.4
  ipv4dnsserver: 130.67.15.198
  ipv6address: 2a02:2121:288:ef9d:bc67:381e:7bf8:f241/64
  ipv6gateway: 2a02:2121:288:ef9d:f446:f8f3:886b:97d0
  ipv6mtu: 1500
  ipv6dnsserver: 2001:4600:4:fff:0:0:0:52
  ipv6dnsserver: 2001:4600:4:1fff:0:0:0:52

```